### DISCUSSION TITLE
How to create a basemodel with relationship?

### DISCUSSION THREAD

**Discussion by PaleNeutron** [fastapi/sqlmodel#858] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from __future__ import annotations
from datetime import datetime
from typing import TYPE_CHECKING, Optional
from sqlalchemy import func

if TYPE_CHECKING:
    from .user import User  # noqa: F401

from sqlmodel import Field, Relationship, SQLModel


class EditableBase:
    create_time: datetime = Field(sa_column_kwargs=dict(server_default=func.now()))
    update_time: datetime = Field(sa_column_kwargs=dict(server_default=func.now(), onupdate=func.now()))
    create_id: int | None = Field(foreign_key="user.id")
    creator: Optional["User"] = Relationship(sa_relationship_kwargs=dict(foreign_keys=[create_id]))
    update_id: int | None = Field(foreign_key="user.id")
    updater: Optional["User"] = Relationship(sa_relationship_kwargs=dict(foreign_keys=[update_id]))
```

### Description

I want to create a base class for every models in my project, all of them need creator and updater.

But if I use it as parent class of my model, this error appears:

```
<class 'app.models.user.User'> has no matching SQLAlchemy type
```

### Operating System

Linux

### SQLModel Version

0.0.16

### Python Version

Python 3.10.12

### Additional Context

this question is different with https://github.com/tiangolo/sqlmodel/issues/488 , My basemodel is not a real table but just a "template"

**Comment by Ringostar1982** [comment]
You mean something like how to inherit ?
Class TokenUser and User will have "first_name" and "last_name" plus the fields declared in its classes too.

```
class BaseUser(SQLModel):
    first_name: str
    last_name: str


class TokenUser(BaseUser):   
    email: EmailStr = Field(default="guest@guest.com")
    locale: Optional[str]
    sub: Optional[str]
    subname: Optional[str]


class User(BaseUser, table=True):    
    __tablename__ = "users"
   division: Optional[str]
   department: Optional[str]

```

**Comment by joachimhuet** [comment, marked as answer]
I think you are just missing the parent class specification: `class EditableBase(SQLModel)`

But from experience it's not a good idea to define relationships in base class, it caused me so many problems.

Now I just put the type and default value for relationships in Base and specify the relationship in Table class. That is where it will actually be used anyway.

Example:

```
class EditableBase(SQLModel):
    create_time: datetime = Field(sa_column_kwargs=dict(server_default=func.now()))
    update_time: datetime = Field(sa_column_kwargs=dict(server_default=func.now(), onupdate=func.now()))

    creator: Optional["User"] = None
    updater: Optional["User"] = None

class EditableTable(EditableBase, table=True):
    # the ids are not used in schema, you can get them from creator.id or editor.id probably
    create_id: int | None = Field(foreign_key="user.id")
    update_id: int | None = Field(foreign_key="user.id")

    # in my experience the foreign_keys does not work properly saying create_id is not a Column so i use `primaryjoin` too
    creator: Optional["User"] = Relationship(sa_relationship_kwargs=dict(primaryjoin="EditableTable.create_id==UserTable.id"))
    updater: Optional["User"] = Relationship(sa_relationship_kwargs=dict(primaryjoin="EditableTable.update_id==UserTable.id"))
```

**Reply by PaleNeutron** [reply]
Thanks, leave a blank creator and updater in parent class is a great work!
